File: src/SymmetricTree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode *root) {
        if (root == NULL)
            return true;
        
        vector <TreeNode*> x, y, temp1, temp2;

        x.push_back(root->left);
        y.push_back(root->right);
        
        while (true) {
            
            int len = x.size();
            if (!len)
                return true;
            for (int i = 0; i < len; i++) {
                if (x[i] == NULL && y[len - i - 1] == NULL)
                    continue;
                if (x[i] == NULL || y[len - i - 1] == NULL)
                    return false;
                if (x[i]->val != y[len - i - 1]->val)
                    return false;
            }
            for (int i = 0; i < len; ++i) {
                if (x[i]) {
                    temp1.push_back(x[i]->left);
                    temp1.push_back(x[i]->right);
                }
                if (y[i]) {
                    temp2.push_back(y[i]->left);
                    temp2.push_back(y[i]->right);
                }
            }
            x = temp1;
            y = temp2;
            temp1.clear();
            temp2.clear();
        }
        return true;
    }
};

File: Algorithms/python/101-SymmetricTree/SymmetricTree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        # empty tree treated ok
        return root is None or self.is_symmetric(root.left, root.right)

    def is_symmetric(self, p, q):
        """
        :param p: first tree
        :param q: second tree
        :return: if p and q is symmetric
        """
        if p is None and q is None:
            return True
        if p is None or q is None:
            return False
        return p.val == q.val and self.is_symmetric(p.left, q.right) and self.is_symmetric(p.right, q.left)
